### PR TITLE
ULFM followup: coverity reports

### PR DESCRIPTION
**Pull request by abouteiller**
There were some new coverity reports in code imported from ULFM. 

This patch set corrects the problematic ones, harmless ones are listed in comments here. 

Note that PRTE errors discovered by the same coverity report have been addressed separately in https://github.com/openpmix/prrte/pull/743

### CODE REVIEW

**Comment by abouteiller** [comment]
________________________________________________________________________________________________________
*** CID 1472862:    (FORWARD_NULL)
/ompi/mca/coll/ftagree/coll_ftagree_earlyreturning.c: 2754 in era_cb_fn()
2748                 result_request(msg_header);
2749                 break;
2750             case MSG_UP:
2751                 msg_up(msg_header, value_bytes, new_dead, ack_failed);
2752                 break;
2753             case MSG_DOWN:
   CID 1472862:    (FORWARD_NULL)
   Passing null pointer "value_bytes" to "msg_down", which dereferences it.
2754                 msg_down(msg_header, value_bytes, new_dead);
2755                 break;
2756             }
2757             opal_mutex_unlock(&era_mutex);
2758         }
2759     
/ompi/mca/coll/ftagree/coll_ftagree_earlyreturning.c: 2751 in era_cb_fn()
2745         else {
2746             switch( msg_header->msg_type ) {
2747             case MSG_RESULT_REQUEST:
2748                 result_request(msg_header);
2749                 break;
2750             case MSG_UP:
   CID 1472862:    (FORWARD_NULL)
   Passing null pointer "value_bytes" to "msg_up", which dereferences it.
2751                 msg_up(msg_header, value_bytes, new_dead, ack_failed);
2752                 break;
2753             case MSG_DOWN:
2754                 msg_down(msg_header, value_bytes, new_dead);
2755                 break;
2756             }
________________________________________________________________________________________________________
*** CID 1472873:  Null pointer dereferences  (FORWARD_NULL)
/ompi/mca/coll/ftagree/coll_ftagree_earlyreturning.c: 2751 in era_cb_fn()
2745         else {
2746             switch( msg_header->msg_type ) {
2747             case MSG_RESULT_REQUEST:
2748                 result_request(msg_header);
2749                 break;
2750             case MSG_UP:
   CID 1472873:  Null pointer dereferences  (FORWARD_NULL)
   Passing null pointer "ack_failed" to "msg_up", which dereferences it.
2751                 msg_up(msg_header, value_bytes, new_dead, ack_failed);
2752                 break;
2753             case MSG_DOWN:
2754                 msg_down(msg_header, value_bytes, new_dead);
2755                 break;
2756             }
_______________________________________________________________________________________________________
*** CID 1472871:    (FORWARD_NULL)
/ompi/mca/coll/ftagree/coll_ftagree_earlyreturning.c: 2751 in era_cb_fn()
2745         else {
2746             switch( msg_header->msg_type ) {
2747             case MSG_RESULT_REQUEST:
2748                 result_request(msg_header);
2749                 break;
2750             case MSG_UP:
   CID 1472871:    (FORWARD_NULL)
   Passing null pointer "new_dead" to "msg_up", which dereferences it.
2751                 msg_up(msg_header, value_bytes, new_dead, ack_failed);
2752                 break;
2753             case MSG_DOWN:
2754                 msg_down(msg_header, value_bytes, new_dead);
2755                 break;
2756             }
/ompi/mca/coll/ftagree/coll_ftagree_earlyreturning.c: 2754 in era_cb_fn()
2748                 result_request(msg_header);
2749                 break;
2750             case MSG_UP:
2751                 msg_up(msg_header, value_bytes, new_dead, ack_failed);
2752                 break;
2753             case MSG_DOWN:
   CID 1472871:    (FORWARD_NULL)
   Passing null pointer "new_dead" to "msg_down", which dereferences it.
2754                 msg_down(msg_header, value_bytes, new_dead);
2755                 break;
2756             }
2757             opal_mutex_unlock(&era_mutex);
2758         }
2759     

THESE ARE INVALID
   because we check msg_header->msglen before derefencing value_bytes in msg_up/msg_down;
   same for new_dead and ack_failed;
   not sure how to deconfuse coverity

________________________________________________________________________________________________________
*** CID 1472875:  Null pointer dereferences  (FORWARD_NULL)
/ompi/mca/coll/ftagree/coll_ftagree_earlyreturning.c: 2394 in result_request()
2388              *  that contextid (modulo rotation on the epoch numbering), and then the
2389              *  number of requested data must be 0, by convention on the last "flushing"
2390              *  agreement that was posted during the free.
2391              */
2392             if( msg_header->agreement_value_header.dt_count == 0 ) {
2393                 /** Then, the answer is "success" */
   CID 1472875:  Null pointer dereferences  (FORWARD_NULL)
   Passing "&success_value" to "send_msg", which dereferences null "success_value.bytes".
2394                 send_msg(NULL, msg_header->src_comm_rank, &msg_header->src_proc_name, msg_header->agreement_id,
2395                          MSG_DOWN, &success_value, 0, NULL);
2396     
2397                 OBJ_DESTRUCT(&success_value);
2398             } else {
2399                 /** Or, I have not started this agreement, or I have started this agreement, but a child

INVALID: success_value.header.dt_count == 0, thus success_value.bytes is never dereferenced.

________________________________________________________________________________________________________
*** CID 1472869:  Null pointer dereferences  (FORWARD_NULL)
/ompi/mca/coll/ftagree/coll_ftagree_earlyreturning.c: 2522 in msg_up()
2516         if( av->header.nb_new_dead > 0 )
2517             av->new_dead_array = new_dead;
2518         else
2519             av->new_dead_array = NULL;
2520     
2521         /* ci holds the current agreement information structure */
   CID 1472869:  Null pointer dereferences  (FORWARD_NULL)
   Passing "av" to "era_combine_agreement_values", which dereferences null "av->new_dead_array".
2522         era_combine_agreement_values(ci, av);
2523         era_update_return_value(ci, msg_header->nb_ack, ack_failed);
2524     
2525         av->new_dead_array = NULL;
2526         av->bytes = NULL;
2527         OBJ_RELEASE(av);

INVALID: value->new_dead_array dereferenced only if nb_new_dead>0 (added an assert)

________________________________________________________________________________________________________
*** CID 1472865:  Null pointer dereferences  (FORWARD_NULL)
/ompi/mca/coll/ftagree/coll_ftagree_earlyreturning.c: 2394 in result_request()
2388              *  that contextid (modulo rotation on the epoch numbering), and then the
2389              *  number of requested data must be 0, by convention on the last "flushing"
2390              *  agreement that was posted during the free.
2391              */
2392             if( msg_header->agreement_value_header.dt_count == 0 ) {
2393                 /** Then, the answer is "success" */
   CID 1472865:  Null pointer dereferences  (FORWARD_NULL)
   Passing "&success_value" to "send_msg", which dereferences null "success_value.new_dead_array".
2394                 send_msg(NULL, msg_header->src_comm_rank, &msg_header->src_proc_name, msg_header->agreement_id,
2395                          MSG_DOWN, &success_value, 0, NULL);
2396     
2397                 OBJ_DESTRUCT(&success_value);
2398             } else {
2399                 /** Or, I have not started this agreement, or I have started this agreement, but a child

INVALID: success_value.new_dead_array referenced only when success_value.msg_header.nb_new_dead > 0

________________________________________________________________________________________________________
*** CID 1472874:    (USE_AFTER_FREE)
/ompi/mca/coll/ftagree/coll_ftagree_earlyreturning.c: 2839 in era_on_comm_rank_failure()
2833                         }
2834                         if( NULL != ci->comm && AGS(comm) != AGS(ci->comm) ) AGS(ci->comm)->ags_status |= AGS_AFR_DIRTY;
2835                         era_mark_process_failed(ci, shadowrank);
2836                     }
2837                 }
2838     
   CID 1472874:    (USE_AFTER_FREE)
   Using freed pointer "next_value".
2839                 value = next_value;
2840             } while( rc == OPAL_SUCCESS );
2841         }
2842     
2843         if( NULL != ompi_stacked_rank_failure_callback_fct ) {
2844             (*ompi_stacked_rank_failure_callback_fct)(comm, rank, remote);
/ompi/mca/coll/ftagree/coll_ftagree_earlyreturning.c: 2839 in era_on_comm_rank_failure()
2833                         }
2834                         if( NULL != ci->comm && AGS(comm) != AGS(ci->comm) ) AGS(ci->comm)->ags_status |= AGS_AFR_DIRTY;
2835                         era_mark_process_failed(ci, shadowrank);
2836                     }
2837                 }
2838     
   CID 1472874:    (USE_AFTER_FREE)
   Using freed pointer "next_value".
2839                 value = next_value;
2840             } while( rc == OPAL_SUCCESS );
2841         }
2842     
2843         if( NULL != ompi_stacked_rank_failure_callback_fct ) {
2844             (*ompi_stacked_rank_failure_callback_fct)(comm, rank, remote);







________________________________________________________________________________________________________
*** CID 1472872:    (TAINTED_SCALAR)
/ompi/dpm/dpm.c: 1605 in ompi_dpm_dyn_init()
1599             ptr = &tmp[1];
1600         } else {
1601             ptr = &tmp[0];
1602         }
1603         port_name = strdup(ptr);
1604     
   CID 1472872:    (TAINTED_SCALAR)
   Passing tainted expression "*port_name" to "ompi_dpm_connect_accept", which uses it as an allocation size.
1605         rc = ompi_dpm_connect_accept(MPI_COMM_WORLD, root, port_name, send_first, &newcomm);
1606         free(port_name);
1607         if (OMPI_SUCCESS != rc) {
1608             return rc;
1609         }
1610     
/ompi/dpm/dpm.c: 1605 in ompi_dpm_dyn_init()
1599             ptr = &tmp[1];
1600         } else {
1601             ptr = &tmp[0];
1602         }
1603         port_name = strdup(ptr);
1604     
   CID 1472872:    (TAINTED_SCALAR)
   Passing tainted expression "*port_name" to "ompi_dpm_connect_accept", which uses it as an allocation size.
1605         rc = ompi_dpm_connect_accept(MPI_COMM_WORLD, root, port_name, send_first, &newcomm);
1606         free(port_name);
1607         if (OMPI_SUCCESS != rc) {
1608             return rc;
1609         }
1610     

I DO NOT UNDERSTAND THIS ERROR 

________________________________________________________________________________________________________
*** CID 1472868:  Null pointer dereferences  (FORWARD_NULL)
/ompi/dpm/dpm.c: 265 in ompi_dpm_connect_accept()
259             goto exit;
260         }
261     
262         /* initiate a list of participants for the connect,
263          * starting with our own members */
264         OBJ_CONSTRUCT(&mlist, opal_list_t);
   CID 1472868:  Null pointer dereferences  (FORWARD_NULL)
   Dereferencing null pointer "members".
265         for (i=0; NULL != members[i]; i++) {
266             OPAL_PMIX_CONVERT_STRING_TO_PROCT(&pxproc, members[i]);
267             plt = OBJ_NEW(opal_proclist_t);
268             memcpy(&plt->procid, &pxproc, sizeof(pmix_proc_t));
269             opal_list_append(&mlist, &plt->super);
270             /* if the rank is wildcard, then we need to skip

THIS IS NOT AN ERROR: We will jump to 'exit' before said dereferencing, and we intentionally do not initialize members so as not to leak them in this error case.




________________________________________________________________________________________________________
To view the defects in Coverity Scan visit, https://u15810271.ct.sendgrid.net/ls/click?upn=HRESupC-2F2Czv4BOaCWWCy7my0P0qcxCbhZ31OYv50yoN-2BQSVjTtaSz8wS4wOr7HlDVSqs52GXSIpR2aUCJKjLn1Fdnx-2FMJPH5oGmXjZMZwc-3D1byv_LK46lj4LWcnuSCmv1wKv6I7b3gGZFEQ68xN5QVZMzCurpSoVJiYjYryO6V0Tk8f7exK5T2cuLQR1GexmXoqlybxsHR9c3RI-2FVLLTdVjGK3Mra2JYRKqEgUmtvVIO0smZNhYg0RS5v06qnR8WweKGPlJCB6TM0ruwx0hDhCOMYZDphjpsHFJOWfzpzLCDZJjxB0mrdz-2BD431ubqk6oUH8-2FOVXjpXoZyVFl-2BfSW2f-2Fo6Y-3D

**Comment by abouteiller** [comment]
Reports that remain uncertain at the moment


```
________________________________________________________________________________________________________
*** CID 1472874:    (USE_AFTER_FREE)
/ompi/mca/coll/ftagree/coll_ftagree_earlyreturning.c: 2839 in era_on_comm_rank_failure()
2833                         }
2834                         if( NULL != ci->comm && AGS(comm) != AGS(ci->comm) ) AGS(ci->comm)->ags_status |= AGS_AFR_DIRTY;
2835                         era_mark_process_failed(ci, shadowrank);
2836                     }
2837                 }
2838     
   CID 1472874:    (USE_AFTER_FREE)
   Using freed pointer "next_value".
2839                 value = next_value;
2840             } while( rc == OPAL_SUCCESS );
2841         }
2842     
2843         if( NULL != ompi_stacked_rank_failure_callback_fct ) {
2844             (*ompi_stacked_rank_failure_callback_fct)(comm, rank, remote);
/ompi/mca/coll/ftagree/coll_ftagree_earlyreturning.c: 2839 in era_on_comm_rank_failure()
2833                         }
2834                         if( NULL != ci->comm && AGS(comm) != AGS(ci->comm) ) AGS(ci->comm)->ags_status |= AGS_AFR_DIRTY;
2835                         era_mark_process_failed(ci, shadowrank);
2836                     }
2837                 }
2838     
   CID 1472874:    (USE_AFTER_FREE)
   Using freed pointer "next_value".
2839                 value = next_value;
2840             } while( rc == OPAL_SUCCESS );
2841         }
2842     
2843         if( NULL != ompi_stacked_rank_failure_callback_fct ) {
2844             (*ompi_stacked_rank_failure_callback_fct)(comm, rank, remote);
```

TODO: I don't see why it's complaining.**




```
________________________________________________________________________________________________________
*** CID 1472872:    (TAINTED_SCALAR)
/ompi/dpm/dpm.c: 1605 in ompi_dpm_dyn_init()
1599             ptr = &tmp[1];
1600         } else {
1601             ptr = &tmp[0];
1602         }
1603         port_name = strdup(ptr);
1604     
   CID 1472872:    (TAINTED_SCALAR)
   Passing tainted expression "*port_name" to "ompi_dpm_connect_accept", which uses it as an allocation size.
1605         rc = ompi_dpm_connect_accept(MPI_COMM_WORLD, root, port_name, send_first, &newcomm);
1606         free(port_name);
1607         if (OMPI_SUCCESS != rc) {
1608             return rc;
1609         }
1610     
/ompi/dpm/dpm.c: 1605 in ompi_dpm_dyn_init()
1599             ptr = &tmp[1];
1600         } else {
1601             ptr = &tmp[0];
1602         }
1603         port_name = strdup(ptr);
1604     
   CID 1472872:    (TAINTED_SCALAR)
   Passing tainted expression "*port_name" to "ompi_dpm_connect_accept", which uses it as an allocation size.
1605         rc = ompi_dpm_connect_accept(MPI_COMM_WORLD, root, port_name, send_first, &newcomm);
1606         free(port_name);
1607         if (OMPI_SUCCESS != rc) {
1608             return rc;
1609         }
1610     
```
Suspect the report stems from using port_name from strdup without first checking for NULL; this has been patched, but we'll need to keep an eye on the next coverity report.

**Review by bosilca — APPROVED** [review]

**Comment by jsquyres** [comment]
@abouteiller Do you have a login to Coverity to see their UI?  They tend to explain in much more detail why each issue was flagged.  Some of the issues can be quite subtle.

If you're 100% sure that issues are moot / Coverity is wrong, you can tell Coverity that it is wrong in their web UI, and then it won't bug us about that issue again.

**Comment by bosilca** [comment]
I was trying to use mine to get access to this data, but it seems it was disabled or I was removed from the project. Any idea ?

**Comment by jsquyres** [comment]
@bosilca @abouteiller I see both of you on the access list in Coverity.  Do you use the github federated login, perchance?

**Comment by awlauria** [comment]
Is this still a WIP or ready to merge?

**Comment by abouteiller** [comment]
It's good to go.

I also marked the reports as false-positive in coverity itself, when appropriate.

**Comment by awlauria** [comment]
@bosilca @abouteiller bring this over to v5.0.x when ready - thanks!

**Comment by gpaulsen** [comment]
@bosilca could you please re-review this, as there have been additional changes since your last review.

File: opal/mca/base/mca_base_var.c [hunk 1]
```diff
@@ -3,7 +3,7 @@
  * Copyright (c) 2004-2008 The Trustees of Indiana University and Indiana
  *                         University Research and Technology
  *                         Corporation.  All rights reserved.
- * Copyright (c) 2004-2012 The University of Tennessee and The University
+ * Copyright (c) 2004-2021 The University of Tennessee and The University
```
Comment (bosilca) [inline opal/mca/base/mca_base_var.c:6]: ?

**Review by bosilca — APPROVED** [review]

**Pull request merged**